doc/overview.md

```diff
@@ -0,0 +1,97 @@
+    1. CoreIRType - a generator can be parameterized by a CoreIR type, which can be used to set the type of the inputs of the modules produced by the generator.
+    1. Module - a generator can accept a CoreIR module as input.
+1. CoreIR Types - These are types of inputs and outputs to circuits. The full list can be seen [here](https://github.com/rdaly525/coreir/blob/master/include/coreir/ir/types.h) if the below list is out of date:
+    1. Bit/BitIn/BitInOut - Bit is for outputs. BitIn is for inputs. BitInOut is for ports that do both. InOuts are typically not used.
+    1. Arrays of CoreIR types
+    1. Records (aka named tuples) of CoreIR types
+    1. NamedTypes that alias other CoreIR types, like aliasing a bit for a CE port.
+
+### Circuits
+1. Param - an object that defines the parameters to a generator.
+    1.
+1. TypeGen - an object that defines the types of the modules produced by generators. Multiple generators can have the same typegen.
+1. Generator Declarations - a declaration of a generator's name, parameters, and typegen.
+1. Generators – functions that produces a module. Parameters to generators should define the structure of the circuit, such as number of inputs or types of variables operated on.
+1. Modules - functions that produce instances. Parameters to modules should not change the structure of the circuit. Rather, they should set things like the constant amount that a counter increments by each clock cycle.
+1. Instances – a circuit in hardware
+    1. Note: in IR, an instance is just a pointer back to a module so that at compilation time, CoreIR knows what to build. This is like LLVM where a call node references a function.
+
+### How To Create And Test A Circuit In CoreIR?
+1. Get a namespace to put the generator in. Aetherling shows how to [make a new namespace](https://github.com/rdaly525/coreir/blob/master/src/libs/aetherlinglib.cpp#L19) and [access it for creating things in it](https://github.com/rdaly525/coreir/blob/master/src/libs/aetherlinglib/aeZip2.h#L10).
+1. Create a params object. See [Aetherling's Zip2 params object](https://github.com/rdaly525/coreir/blob/master/src/libs/aetherlinglib/aeZip2.h#L18) for an example of how to create one.
+1. Create a typegen. See [Aetherling's Zip2 typegen object](https://github.com/rdaly525/coreir/blob/master/src/libs/aetherlinglib/aeZip2.h#L25) for an example.
+    1. Note that the record returned by the anonymous function is the type of the modules produced by the generators that use this typegen.
+1. Create a generator declaration. See [Aetherling's Zip2 generator](https://github.com/rdaly525/coreir/blob/master/src/libs/aetherlinglib/aeZip2.h#L46) for an example.
+1. Add a unit test that just verifies the generator can instantiate modules. This is a basic test that doesn't guarantee correctness. Do this by creating a .cpp file in the tests/unit folder and following [simple.cpp's example](https://github.com/rdaly525/coreir/blob/master/tests/unit/simple.cpp#L6)
```

Comment:
"Correctness" is dependent on what you mean by correct. For a compiler, correctness means that the generated circuit is a valid circuit that is functionally the same as what was described. For a front end/simulator, correctness means that the circuit does what the test expects it to do.